fpga_interchange/site_lut_mapping_cache.h

```diff
@@ -0,0 +1,110 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2021  Symbiflow Authors
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#ifndef SITE_LUT_MAPPING_CACHE_H
+#define SITE_LUT_MAPPING_CACHE_H
+
+#include "idstring.h"
+#include "nextpnr_namespaces.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+// Key structure used in site LUT mapping cache
+struct SiteLutMappingKey {
+
+    // LUT Cell data
+    struct Cell {
+        IdString                    type;       // Cell type
+        int32_t                     belIndex;   // Bound BEL index
+        dict<IdString, IdString>    conns;      // Port net assignments
+
+        bool operator == (const Cell& other) const {
+            return  (type == other.type) &&
+                    (belIndex == other.belIndex) &&
+                    (conns == other.conns);
+        }
+
+        bool operator < (const Cell& other) const {            
+            return belIndex < other.belIndex;
+        }
+    };
+
+    int32_t           tileType;
+    int32_t           site;
+    std::vector<Cell> cells;
+
+    static SiteLutMappingKey create (const SiteInformation& siteInfo);
+   
+    bool operator == (const SiteLutMappingKey &other) const {
+        return (tileType == other.tileType) &&
+               (site == other.site) &&
+               (cells == other.cells);
+    }
+
+    bool operator != (const SiteLutMappingKey &other) const {
+        return (tileType != other.tileType) ||
+               (site != other.site) ||
+               (cells != other.cells);
+    }
+
+    unsigned int hash () const {
```

Comment:
if the key is immutable, it might be more efficient to compute the hash once when creating the key, then store it and only have to return it when `hash()` is called